src/package-hoister.js

```diff
@@ -196,7 +200,12 @@ export default class PackageHoister {
 
       for (const depPattern of ref.dependencies) {
         const depinfo = this._lookupDependency(info, depPattern);
-        if (depinfo && !depinfo.isRequired && !depinfo.isIncompatible) {
+        if (
+          depinfo &&
+          !depinfo.isRequired &&
+          !depinfo.isIncompatible &&
+          (!depinfo.pkg._reference || !depinfo.pkg._reference.optional || !this.ignoreOptional)
```

Comment:
Yeah I agree, but there was some subtle changes in the conditions so I aimed to make the minimal change required and revisit if needed. I'll give it a second look.